source/fab/builder.py

```diff
@@ -221,7 +222,16 @@ def run(self, source: Path):
             #       expect to be *produced* by the compile
             #       and pass this to the constructor for
             #       inclusion in the task's "products"
-            compiler_class = self._compiler_map[unit.found_in.suffix]
+            compiler_class = None
+            for pattern, classname in self._compile_map:
+                # Note we keep searching through the map
+                # even after a match is found; this means that
+                # later matches will override earlier ones
+                if match(pattern, str(unit.found_in)):
```

Comment:
I note the match is being done from text each time. This will be considerably slower than using a compiler pattern however that can be an optimisation for later.

Reply:
One thing to keep in mind (and related to the above comment too) is that (eventually, in the wider context of what this is going towards with the "phases") the patterns end up containing substitutions as well.  e.g. the pattern can contain `{source}` which gets substituted for the source path of the current phase (and this happens at the start of each phase).  But perhaps that is the perfect place to eventually stick a call to `compile` the pattern (it may need an intermediate object for that, but maybe that can be part of the "phase" / "map" object discussed above

Reply:
It may be that the subsitution becomes unecessary down the line but even if we can just compile and use twice we win over two uses.